### PR TITLE
Sv/local matfree scpccg rebased150522

### PR DESCRIPTION
**Pull request by dham**

### CODE REVIEW

File: firedrake/slate/slac/compiler.py [hunk 1]
```diff
@@ -642,36 +668,59 @@ def parenthesize(arg, prec=None, parent=None):
...
-    assignments = list(zip(ret_vars, preprocessed_gem_expr))
+    assignments = list(zip([gem.Indexed(gem.Variable(out_name, shape), idx)], preprocessed_gem_expr))
 
     # Part A: slate to impero_c
     impero_c = impero_utils.compile_gem(assignments, (), remove_zeros=False)
 
     # Part B: impero_c to loopy
     output_arg = OutputKernelArg(output_loopy_arg)
-    return generate_loopy(impero_c, args, scalar_type, "slate_loopy", [], log=PETSc.Log.isActive()), output_arg
+    knl_name = knl_prefix + "_knl_%d" % knl_counter()  # auto generate a unique name
```
Comment (sv2518) [inline firedrake/slate/slac/compiler.py:721]: Try to use loopy naming infrastructure, note that naming is for the function inside the file and only needs to be unique to the file

File: firedrake/slate/slac/kernel_builder.py [hunk 2]
```diff
@@ -434,17 +444,18 @@ def __init__(self, expression, tsfc_parameters=None):
...
+        self.bag = SlateWrapperBag({}, coords=coords)
+        self.matfree_solve_knls = []
 
     def tsfc_cxt_kernels(self, terminal):
         r"""Gathers all :class:`~.ContextKernel`\s containing all TSFC kernels,
         and integral type information.
         """
 
-        return compile_terminal_form(terminal, prefix=f"subkernel{next(self.kernel_counter)}_",
+        return compile_terminal_form(terminal, prefix="subkernel%d_" % knl_counter(),
```
Comment (sv2518) [inline firedrake/slate/slac/kernel_builder.py:458]: see previous comment

File: firedrake/slate/slac/kernel_builder.py [hunk 3]
```diff
@@ -505,10 +537,51 @@ def loopify_tsfc_kernel_data(self, kernel_data):
...
+                    # reset offset when all splits of one mixed coefficient have been dealt with
+                    # and a new mixed coefficient in kernel data is dealt with
+                    # NOTE this depends on all split coefficients being appended in order
+                    # (which is the case but noteworthy nevertheless)
+                    offset = 0
+                    last_mixed_c = None
+
+                shp2 = self.extent(split_c)
+                shp, = shp2 if shp2 else (1,)
+                idx = self.bag.index_creator((shp,))
```
Comment (dham) [inline firedrake/slate/slac/kernel_builder.py:570]: `shp, = self.extent(split_c) or (1,)`

File: firedrake/slate/slac/kernel_builder.py [hunk 4]
```diff
@@ -1,27 +1,37 @@
...
 
 from functools import singledispatch
 import firedrake.slate.slate as slate
 from firedrake.slate.slac.tsfc_driver import compile_terminal_form
+from firedrake.slate.slac.kernel_settings import knl_counter
 
 from tsfc import kernel_args
 from tsfc.finatinterface import create_element
 from tsfc.loopy import create_domains, assign_dtypes
+from pyop2 import configuration
```
Comment (JDBetteridge) [inline firedrake/slate/slac/kernel_builder.py:27]: Remove this import, you don't seem to use it anywhere

File: firedrake/slate/slac/kernel_builder.py [hunk 5]
```diff
@@ -517,35 +590,35 @@ def layer_integral_predicates(self, tensor, integral_type):
...
         i = self.bag.index_creator((1,))
-        predicates = [pym.Comparison(pym.Subscript(pym.Variable(self.cell_facets_arg_name), (fidx[0], 0)), "==", select)]
+        predicates = [str(pym.Comparison(pym.Subscript(pym.Variable(self.cell_facets_arg_name), (fidx[0], 0)), "==", select))]
 
         # TODO subdomain boundary integrals, this does the wrong thing for integrals like f*ds + g*ds(1)
         # "otherwise" is treated incorrectly as "everywhere"
         # However, this replicates an existing slate bug.
         if kinfo.subdomain_id != "otherwise":
-            predicates.append(pym.Comparison(pym.Subscript(pym.Variable(self.cell_facets_arg_name), (fidx[0], 1)), "==", kinfo.subdomain_id))
+            predicates.append(str(pym.Comparison(pym.Subscript(pym.Variable(self.cell_facets_arg_name), (fidx[0], 1)), "==", kinfo.subdomain_id)))
```
Comment (dham) [inline firedrake/slate/slac/kernel_builder.py:621]: Is this still needed?

File: firedrake/slate/slac/kernel_builder.py [hunk 6]
```diff
@@ -572,34 +645,72 @@ def is_integral_type(self, integral_type, type):
...
+                new = True
+            except (KeyError, TypeError):
+                # if coefficient is not in names it is not an
+                # an action coefficient so we can use usual naming conventions
+                if not new:
+                    if c not in coeff_dict.keys():
+                        count = i+len(coeff_dict)
+                    else:
+                        count = list(coeff_dict.keys()).index(c)
+                        count2 = len(coeff_dict[c])
```
Comment (dham) [inline firedrake/slate/slac/kernel_builder.py:678]: write only variable?

File: firedrake/slate/slac/kernel_builder.py [hunk 7]
```diff
@@ -572,34 +645,72 @@ def is_integral_type(self, integral_type, type):
         else:
             return False
 
-    def collect_coefficients(self):
-        """Saves all coefficients of self.expression where non-mixed coefficients
-        are dicts of form {coeff: (name, extent)} and mixed coefficients are
-        double dicts of form {mixed_coeff: {coeff_per_space: (name, extent)}}.
+    def collect_coefficients(self, expr=None, names=None, artificial=True):
```
Comment (sv2518) [inline firedrake/slate/slac/kernel_builder.py:648]: Add docs on what the args are.

File: firedrake/slate/slac/kernel_builder.py [hunk 8]
```diff
@@ -654,61 +769,307 @@ def slate_call(self, prg, temporaries):
...
+        name = "mtf_solve_%d" % knl_no
+        shape = expr.shape
+        dtype = self.tsfc_parameters["scalar_type"]
+        rtol = getattr(expr.ctx, "rtol")
+        atol = getattr(expr.ctx, "atol")
+        prec = getattr(expr.ctx, "preconditioner")
+        max_it = int(getattr(expr.ctx, "max_it")) if getattr(expr.ctx, "max_it") else 2*shape[0]
+        preconditioned = bool(prec)
+
+        # if max_it:
```
Comment (dham) [inline firedrake/slate/slac/kernel_builder.py:789]: Dead code

File: firedrake/slate/slac/kernel_builder.py [hunk 9]
```diff
@@ -654,61 +769,307 @@ def slate_call(self, prg, temporaries):
...
+             stop_criterion,
+             f"""
+                    <> beta = rkp1_norm / rk_norm {{dep={stop_criterion_id}, id=beta}}
+                    rk_norm = rkp1_norm {{dep=beta, id=rk_normk}}
+                    {p}[i_10] = beta * {p}[i_10] - {z}[i_10] {{dep=rk_normk, id=projectork}}
+                    {A_on_p}[i_12] = 0. {{dep=projectork, id=Aonp0, inames=i_c}}
+                    max_it_crit = i_c {{dep=Aonp0, id=tempk, inames=i_c}}
+                  end""",
+            loopy.CInstruction("",
+                                  f"{{PetscPrintf(PETSC_COMM_WORLD, \"The local solver {name} has not run to convergence.\"); PetscFinalize(); exit(1);}}",
```
Comment (dham) [inline firedrake/slate/slac/kernel_builder.py:885]: Can't be on `PETSC_COMM_WORLD`.

File: firedrake/slate/slac/kernel_builder.py [hunk 10]
```diff
@@ -654,61 +769,307 @@ def slate_call(self, prg, temporaries):
...
+                                  code="break;",
+                                  predicates={parse(f"rtol_crit < {rtol} or atol_crit < {atol}")},
+                                  depends_on=dep,
+                                  id=id)
+
+    def generate_kernel_args_and_call_reads(self, expr, insn, dtype):
+        """A function which is used for generating the arguments to the kernel and the read variables
+           for the call to that kernel of the matrix-free solve.
+
+           TODO Add comment on the order of the arguments."""
```
Comment (JDBetteridge) [inline firedrake/slate/slac/kernel_builder.py:964]: TODO

File: firedrake/slate/slac/kernel_builder.py [hunk 11]
```diff
@@ -760,26 +1121,51 @@ def generate_tsfc_calls(self, terminal, loopy_tensor):
...
         self.needs_mesh_layers = False
-        self.call_name_generator = UniqueNameGenerator(forced_prefix="tsfc_kernel_call_")
+        self.num_facets = None
+        self.call_name_generator = UniqueNameGenerator()
         self.index_creator = IndexCreator()
+        self.name = name
+        self.coords = coords
+
+    def copy_extra_args(self, other):
+        # self.coefficients.update({c:k for c, k in other.coefficients.items() if c not in self.coefficients})
```
Comment (JDBetteridge) [inline firedrake/slate/slac/kernel_builder.py:1147]: dead code?

File: firedrake/slate/slac/kernel_builder.py [hunk 12]
```diff
@@ -760,26 +1121,51 @@ def generate_tsfc_calls(self, terminal, loopy_tensor):
...
+        if not self.needs_cell_sizes:
+            self.needs_cell_sizes = other.needs_cell_sizes
+        if not self.needs_cell_facets:
+            self.needs_cell_facets = other.needs_cell_facets
+            self.num_facets = other.num_facets
+        if not self.needs_mesh_layers:
+            self.needs_mesh_layers = other.needs_mesh_layers
+
+    def copy_coefficients(self, coeffs=None, action_coeffs=None):
+        # if coeffs:
```
Comment (JDBetteridge) [inline firedrake/slate/slac/kernel_builder.py:1159]: dead code?

**Pull request closed without merging**